pagopa/pn-frontend
# Pn 2566 fe recupero iun e mandate id partendo da link rapido di aar

## Short description
This PR adds a component linked to the path `/notifiche/qrcode/:qrcode` that gets the id for the notification from the value of the path param `qrcode`, and then redirects to the notification detail given the obtained id.  
The QR code is generated by a separated process (I guess related to the "Avviso di Avvenuta Ricezione", thus the "aar" in the PR title) which refers to a particular recipient in case of a multirecipient notification.  

The notification id is obtained through an API call, which takes the logged user from the auth header. The corresponding endpoint was recently developed by Alessandro Masci.
The id can be just `{ iun }` if the logged user is the destinated recipient, or `{ iun, mandateId }` if the logged user is a delegate of the destinated recipient. For any other user, the call answers with a 404 status. In this case, the FE shows a blank page with a message 'Link alla notifica non trovato" and a button to proceed to the home page.
We checked the endpoint for the recipient, a delegate, and other user, obtaining in all cases the expected response. Both the 106-char and the 136-char QR codes were tested, based on an example case given by Alessandro.   
As the modification done recently by (I guess) Lorenzo Cocciarelli, which preserves the requested path after the login, is already integrated; the link with the QR code can be used to go to the login, and then immediately to the corresponding notification detail.

Of course, this PR affects the functionality of pf-webapp only.

## List of changes proposed in this pull request
We added the `NotificationFromQrCode` component, it launches the "QR code exchange" and then performs the redirect to `/notifiche/:iun/dettaglio` or `/notifiche/:mandateId/:iun/dettaglio`.  
We added the endpoint for the "QR code exchange" to `NotificationApi`.

Two little adaptations to pn-commons were needed.
- We added the optional `message` and `subtitle` props to `AccessDenied`. If present, the values of such props take precedence over the default message and subtitle shown by this component.
- We added the optional `showBackAction` prop, with default `true`, to `PnBreadcrumb`. If set to `false`, then the "Indietro" link (cfr. the image below) is not shown.

![image](https://user-images.githubusercontent.com/5519535/203072832-af41d4a2-0862-4198-8602-cd3bc637ae3d.png)

A little bug fix is also included: the "Notifiche" link in the same breadcrumb for the notification detail page, if the logged user (say A) is a delegate for a recipient (say B), points to the list of notifications for that recipient (B in the example). Previously this link pointed to the list of notifications for the logged user as recipient (the notifications having A as recipient in the example).

### NB
We included tests for (I hope) all the additions and modifications included in this PR.


## How to test
For the DEV environment, the following links point to a notification whose only recipient is Ettore Fieramosca.
(106 chars)
https://portale.dev.pn.pagopa.it/notifiche/qrcode/WlhQSC1ZVExBLUFOVEctMjAyMjEwLVItMV9GUk1UVFI3Nk0wNkI3MTVFX2ZkODZlNmFjLWUzNDktNDRhZS1hODFhLWVjZGRiOTEzOTJiOQ
(136 chars)
https://portale.dev.pn.pagopa.it/notifiche/qrcode/VVFOUC1VV0pBLUVEWkctMjAyMjExLUgtMV9QRi00ZmM3NWRmMy0wOTEzLTQwN2UtYmRhYS1lNTAzMjk3MDhiN2RfZDc1NjM1OGYtMzI1Yi00ZjY2LTgzMjQtZTM4ZjljMGU1MTRk

In turn, Cristoforo Colombo has an active delegation that allows him to see the notifications for Ettore Fieramosca.

Hence, the preceding links can be used to test the three scenarios: recipient, delegate, and rejected request (e.g. if the link is used by Ada Lovelace).
